### ISSUE TITLE
Test sharding / parallelization

### ISSUE THREAD

**Issue by vojtajina** [karma-runner/karma#439]
Allow splitting a single test suite into multiple chunks, so that it can be executed in multiple browsers and therefore executed on multiple machines or at least using multiple cores.

**Comment by jbnicolai** [comment]
@EtaiG I have not started working on this, but as my project currently exceeds 3000 tests as well it's becoming something I want to invest some time into as well.

**Comment by danyshaanan** [comment]
A note about dynamic creation of groups (@vojtajina) - We should be aware of how this affects tests that happen to have side effects. Imagine tests B, C, and D, and a naive alphabetical devision of tests into two groups - {B,C} and {D}. Lets say C has a devastating side affect and I'm adding test A, hence changing the grouping to {A,B} and {C,D}. Now D will fail, just because the grouping changed.

Of course, tests shouldn't have side effects, but this case is bound to happen, and might be very confusing to users.

**Comment by EtaiG** [comment]
I think we can ignore this case and let people who encounter it deal with
their own problems.
We can also enable the API for this to allow the consumer to decide the
grouping.

**Comment by scriby** [comment]
+1

We have a large number of tests at work, and sharding would be very beneficial. As you said, there shouldn't be side effects between tests, and for anyone who doesn't want to remove the side effects, I'd say they just don't get to run their tests in parallel :)

As long as the sharding is opt-in I think the confusion should be manageable.

**Comment by LFDM** [comment]
Hey @EtaiG and @danyshaanan, very interested in this experiment you mentioned. Is this code accessible somewhere? I'd very much like to experiment with this a bit - maybe your work could give me a headstart!

**Comment by danyshaanan** [comment]
@LFDM We have nothing to share at the moment, but I'm just about to rewrite a smarter version of it in the coming couple of weeks. I'll try to do so in a way I'll be able to share. Feel free to ping me about this in a week or so if I'll not post anything by then.

**Comment by AlanFoster** [comment]
:+1: Sounds like a great idea to me; Would love to see any progress updates on this @danyshaanan!

**Comment by park9140** [comment]
:+1:  this would be great.

**Comment by aaichlmayr** [comment]
@danyshaanan! any news?

**Comment by danyshaanan** [comment]
@aaichlmayr :
yeah, bad ones - It didn't seem to work out that well; This feature's spec is non trivial, therefore the implementation was not as clean and I would hope, and the benefits were not convincing enough to go ahead with it, so we scrapped this plan.

**Comment by aaichlmayr** [comment]
Thanks for the info

**Comment by booleanbetrayal** [comment]
> As long as the sharding is opt-in I think the confusion should be manageable.

Totally agree. Think it's fine to launch as an experimental feature with this requirement. Would love to see this land and would be happy to help bug-hunt, etc.

**Comment by FezVrasta** [comment]
Hi, has the feature been shipped already?

**Comment by navels** [comment]
:+1: anyone making headway on this?

**Comment by presidenten** [comment]
So is this a feature yet?

**Comment by presidenten** [comment]
I dont get it? Did I do somethink wrong? What did I miss? Why the thumbs down?

**Comment by FezVrasta** [comment]
If you reply to an issue, all the subscribed people get an email and a notification. If you just want to add a +1 on the issue, do so adding a thumb up reaction to the first post (or to the one with more upvotes), in this way you don't flood the whole list of subscribers.

**Comment by Florian-R** [comment]
@dignifiedquire Could you lock this one like #1320 with a help:wanted label? Thanks!

**Comment by pauldraper** [comment]
> When using a dependency management system this will be much easier and simplified

True, but it's a hack to get these systems to work in karma in the first place right? I'm tempted to put that consideration aside for now.

I agree with the suggestions made by @vojtajina https://github.com/karma-runner/karma/issues/439#issuecomment-20945728 (even if they are three and a half years old :)

I thinking

```javascript
module.exports = function(config) {
  config.set({
    files: [
      'lib/lib1.js',
      'lib/lib2.js',
      {'other/**/*.js', included: false},
      {pattern: 'app/**/*.js', grouped: true},
      {patttern: 'more-app/**/*.js', grouped: true}, 
    ],
  });
}
```

And then the resolved object would be

```javascript
{
  served: [
    'other/file1.js',
    'other/files/file2.js'
  ],
  included: {
    common: [
      'lib/lib1.js',
      'lib/lib2.js'
    ]
    groups: {
      0: {
        'app/file1.js',
        'app/file2.js',
      },
      1: {
        'app/file3.js',
        'more-app/file.js',
      }
    }
  }
}
```

We could reuse `concurrency`, though a default of `Infinity` is bad -- most commonly we want to run as many tests as we have cores.

**Comment by habermeier** [comment]
I'd hate to have to group tests by hand.  What if we had the system uses the regular configuration as a starting point, and have it build up and refine an optimal parallel test plan over time?  Along the way it might be able to discover any dependency chains (that shouldn't be there, but might be).  It could flag those as "todo" items for developers, but could work around those should it discover them.  Whatever it does, it'd be good for it to be able to deal with changes in the test code gracefully, so it would not have to recompute the whole thing all over again when a single test is added (or removed).

I'm sure the computational complexity would be enormous for getting at the very best configuration, but maybe some rough heuristics would get us reasonably close.

**Comment by FezVrasta** [comment]
really, just copy what jest does. it's fine

**Comment by pauldraper** [comment]
> I'd hate to have to group tests by hand.

Not sure if I understand this right, but I wasn't suggesting that.

```
{pattern: 'app/**/*.js', grouped: true},
{patttern: 'more-app/**/*.js', grouped: true}, 
```

`grouped` just means "these are the files I want to split into different test groups. (As opposed to the common library files that are in *all* the tests groups.)

Karma then generates a number of arbitrary groups automatically from those locations. I suggest a `groups` config options for the number of groups. It can be tuned to weigh scheduling efficiency against startup overhead.

**Comment by brandonros** [comment]
Is this dead in the water?

**Comment by vikerman** [comment]
Hello - Here is my proposal for running tests in parallel.

This is a very simple sharding strategy - But it should provide speedup just by using multiple processors on the machine. This is meant mostly for local development and not much for CI runs(where remote CI setup costs far outweigh the speed gains of parallelization)

karma.js changes:
- The root Karma URL can take in 3 extra URL parameters - shardId, shardIndex, totalShards
- This is passed on to the context iframe

Jasmine(or Mocha etc.) Adapter changes:
- In the context page, the Jasmine(or mocha) adapter can process the shardIndex and totalShards parameters
- The adapter passes the  shardId, shardIndex and totalShards in the "info" object it uses while connecting back to the Karma sever(See section below on how it's used) 
- The adapter walks the suite/spec tree and collects all leaf specs in an array
- The adapter uses a very simple sharding strategy - It runs the subset of tests 
   [(shardIndex/totalShards * totalTests) -> ((shardIndex + 1)/totalShards * totalTests)]

Karma server changes:
- The server now has logic to wait for all shards to connect before starting a test run
- This is so that the test execution doesn't immediately start when the browser instance corresponding to the first shard connects and run once more when the rest of the shards connect.
- Server uses "shardId" and "totalShards" to determine whether enough number of sharded browsers have connected with the same "shardId"

Chrome(or other launcher) changes:
- "ChromeSharded" (and "ChromeHeadlesSharded etc.) is a new type of launcher that launches Chrome with N different tabs - each with the same shardId and totalShards and the appropriate shardIndex
- Default number of shards  = Number of processors on the machine
- Overriden by some Launcher flag / Environment variable

Reporter changes: (I need to flush this out more. Any ideas welcome here)
- No changes in initial version - For local runs reporter output doesn't matter? Prints error from any of the shards on the console
- Ideally reporter can collate all results from all the shards into a single report

**Comment by brandonros** [comment]
I'm going to try to get this done this weekend. I don't know anything about this project or the codebase, but I think a ton of people would be saved a ton of time if I can figure something out. Maybe different tabs running on different ports? Could get hairy...

**Comment by littleninja** [comment]
@brandonros rock on! I'm in a similar boat--haven't contributed to the project before but would also be interested if I can help with a sharding feature. Would you be willing to create a fork to get the ball rolling? I don't know--do people usually create "WIP - xyz feature" pull request to help rally effort?

**Comment by brandonros** [comment]
Sarah,

Let's link up and see where we can hook into this and dive in. When are you free? (probably best not to continue this conversation here)

Regards,

Brandon
________________________________

**Comment by brandonros** [comment]
So, here's what happens at a high level:

1. your package calls Karma with a configuration describing a framework like Jasmine

2. A browser is opened and pointed to the local Karma server, which serves up some HTML/JS listing your files to test, and an iframe, then hands it off to the framework

I actually think it would make more sense to inject into Jasmine, because I was able to boil down exactly where the tests are executed. However, I ran into an issue where it didn't like that I was trying to execute different suites at the same time.

So, I came up with `http://localhost:9876/?shardIndex=0&numShards=4` at https://github.com/brandonros/karma

It is still very early. All it does is allow you to chunk the test files (I hardcoded that they have to contain Spec) across multiple tabs. I'll try to update if I make any worthwhile progress on a more complete package.

**Comment by rschuft** [comment]
I created a plugin that automatically shards tests across the listed browsers (e.g. if you want two sets you list two browsers... browsers: ['ChromeHeadless', 'ChromeHeadless']).  It doesn't achieve one of the concerns listed in this thread: run tests at the same time.  It forces concurrency: 1.  It does however fix the memory problems of having too many specs loading in a single browser and it correctly works with coverage reporting.

[karma-sharding](https://www.npmjs.com/package/karma-sharding)

**Comment by joeljeske** [comment]
I have also created a plugin, [karma-parallel](https://www.npmjs.com/package/karma-parallel) similar to the one that @rschuft made but is a bit different. It supports running specs in different browser instances by splitting up the commonly used `describe` as opposed to splitting up the spec files themselves. This allows you to run tests in parallel, even after using a bundler such as karma-webpack or karma-browserify. It also supports testing in multiple types of browsers. 

[karma-parallel](https://www.npmjs.com/package/karma-parallel)

**Comment by guilhermejcgois** [comment]
@joeljeske do u have tested with angular project using ng-cli? we have one and your plugin seems to be interesting to us, will give a try

**Comment by joeljeske** [comment]
I do not use an ng-cli project on a regular basis, but I have done basic testing on an ng-cli project and it works just fine. Please log any issues if you run into to something. 

One note; it is not yet tested with coverage reporting. It would likely be best to disable coverage reporting when using karma-parallel. Coverage reporting is a future feature of the plugin.

**Comment by rschuft** [comment]
The karma-sharding plugin doesn't play well with the ng-cli because of the way webpack bundles the specs together before the middleware engages with it.  Hopefully @joeljeske can bypass that limitation with his approach.

**Comment by joeljeske** [comment]
@guilhermejcgois, the latest release of [karma-parallel](https://www.npmjs.com/package/karma-parallel) is tested and compliant with ng-cli projects.  Code coverage support was just introduced. Would love to hear feedback on your experience with it.

**Comment by intellix** [comment]
My results from karma-parallel on MBP 15" Early 2013 (8 cores): https://github.com/joeljeske/karma-parallel/issues/1#issuecomment-393447431

```
1x --------------
real 1m25.659s
user 1m42.960s
sys 0m10.470s

2x --------------
real 1m15.996s
user 2m1.312s
sys 0m13.601s

3x --------------
real 1m11.746s
user 2m23.417s
sys 0m18.218s

4x --------------
real 1m13.970s
user 2m36.800s
sys 0m19.698s

7x just timed out
```

Was expecting at least 2x perf but seems not to really make a difference

**Comment by wbhob** [comment]
I'm having an issue where, when one shard disconnects, it simply stops that group of tests and moves on to the next – then, it returns exit code 0. Any thoughts on what may be causing this?

Karma ^3.1.4
Jasmine ^2.99
karma-parallel ^0.3.1

**Comment by rschuft** [comment]
I’m not sure.  It sounds like it is not really collecting the failures at all.  Which browser are you firing off?

**Comment by wbhob** [comment]
I’ve tried Chrome and ChromeHeadless

**Comment by juanmendes** [comment]
> @guilhermejcgois, the latest release of [karma-parallel](https://www.npmjs.com/package/karma-parallel) is tested and compliant with ng-cli projects. Code coverage support was just introduced. Would love to hear feedback on your experience with it.

We use karma-parallel and we're getting false positives because one of the shards isn't running all the tests. There are test problems, but karma-parallel hides them because it doesn't fail the tests. After a browser disconnects, typically a shard is restarted from the beginning. However, s sometimes they restart and run just one of the tests. It definitely has to do with the size of the project. We're running 3500 tests where over 200 of them are ng component DOM testing. I've created an issue here https://github.com/joeljeske/karma-parallel/issues/42
